src/lib/propagation/DAG.js

```diff
@@ -0,0 +1,63 @@
+import * as NodeFns from './nodeFns'
+import {SimulationNode} from './node'
+import * as constants from './constants'
+
+import * as _collections from 'gEngine/collections'
+
+export class SimulationDAG {
+  constructor(nodes) {
+    if (!!_.get(window, 'recorder')) { window.recorder.recordSimulationDAGConstructionStart(this) }
+    let rest = nodes.map(NodeFns.extractInputs)
+
+    let byId = _.transform(rest, (map, node) => {map[node.id] = {node, lastAncestors: node.inputs, ancestors: node.inputs}}, {})
+
+    const missingInputsNodes = _.remove(rest, NodeFns.hasMissingInputs(rest))
+    let graphErrorNodes = missingInputsNodes.map(NodeFns.withMissingInputError(nodes))
+    let errorNodes = Object.assign([], graphErrorNodes)
+    let heightOrderedNodes = []
+
+    while (!_.isEmpty(rest)) {
+      const nextLevelNodes = _.remove(
+        rest,
+        n => NodeFns.allInputsWithin(heightOrderedNodes)(n) && _.isEmpty(n.errors) && !NodeFns.anyInputsWithin(errorNodes)(n)
+      )
+      heightOrderedNodes.push(...nextLevelNodes)
+
+      const incomingErrorNodes = _.remove(rest, n => !_.isEmpty(n.errors) && NodeFns.allInputsWithin(heightOrderedNodes)(n))
+      errorNodes.push(...incomingErrorNodes)
+      heightOrderedNodes.push(...incomingErrorNodes) // We may want to resimulate these later anyways...
+
+      const infiniteLoopNodes = _.remove(rest, n => _.some(byId[n.id].lastAncestors, id => id === n.id))
+      const withInfiniteLoopErrors = infiniteLoopNodes.map(NodeFns.withInfiniteLoopError)
+      errorNodes.push(...withInfiniteLoopErrors)
+      graphErrorNodes.push(...withInfiniteLoopErrors)
+
+      const inputErrorNodes = _.remove(
+        rest,
+        _collections.andFns(NodeFns.anyInputsWithin(errorNodes), NodeFns.allInputsWithin([...heightOrderedNodes, ...errorNodes]))
+      )
+      const withAncestralErrors = inputErrorNodes.map(NodeFns.withAncestralError(errorNodes))
+      errorNodes.push(...withAncestralErrors)
+      graphErrorNodes.push(...withAncestralErrors)
+
```

Comment:
I bet there's a cleaner way to do all of this (though I realize it is much cleaner than it could be, and this stuff is complex).  Having one object/class for the collection could make this very neat:
``DAGnodes.sortErrors``, then in ``sortErrors``: 
``{goodNodes, badNodes} = DAGnodes.checkIncomingErrors.checkInfiniteLoopErrors.checkInputErrors``